nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/PutDatabaseRecord.java

```diff
@@ -0,0 +1,1058 @@
+
+            // Get the statement type from the attribute if necessary
+            String statementType = statementTypeProperty;
+            if (USE_ATTR_TYPE.equals(statementTypeProperty)) {
+                statementType = flowFile.getAttribute(STATEMENT_TYPE_ATTRIBUTE);
+            }
+            if (StringUtils.isEmpty(statementType)) {
+                log.error("Statement Type is not specified, flowfile {} will be penalized and routed to failure", new Object[]{flowFile});
+                flowFile = session.penalize(flowFile);
+                session.transfer(flowFile, REL_FAILURE);
+            } else {
+                RecordSchema recordSchema;
+                try (final InputStream in = session.read(flowFile)) {
+
+                    final RecordReader recordParser = recordParserFactory.createRecordReader(flowFile, in, log);
+                    recordSchema = recordParser.getSchema();
+
+                    if (SQL_TYPE.equalsIgnoreCase(statementType)) {
+
+                        // Find which field has the SQL statement in it
+                        final String sqlField = context.getProperty(FIELD_CONTAINING_SQL).evaluateAttributeExpressions(flowFile).getValue();
+                        if (StringUtils.isEmpty(sqlField)) {
+                            log.error("SQL specified as Statement Type but no Field Containing SQL was found, flowfile {} will be penalized and routed to failure", new Object[]{flowFile});
+                            flowFile = session.penalize(flowFile);
+                            session.transfer(flowFile, REL_FAILURE);
+                        } else {
+                            boolean schemaHasSqlField = recordSchema.getFields().stream().anyMatch((field) -> sqlField.equals(field.getFieldName()));
+                            if (schemaHasSqlField) {
+                                try (Statement s = con.createStatement()) {
+
+                                    try {
+                                        s.setQueryTimeout(queryTimeout); // timeout in seconds
+                                    } catch (SQLException se) {
+                                        // If the driver doesn't support query timeout, then assume it is "infinite". Allow a timeout of zero only
+                                        if (queryTimeout > 0) {
+                                            throw se;
+                                        }
+                                    }
+
+                                    Record currentRecord;
+                                    while ((currentRecord = recordParser.nextRecord()) != null) {
+                                        Object sql = currentRecord.getValue(sqlField);
+                                        if (sql != null && !StringUtils.isEmpty((String) sql)) {
+                                            // Execute the statement as-is
+                                            s.execute((String) sql);
```

Comment:
I agree that it should be configurable by the user whether to include begin and commit directives. For step 5 above, the "begin" event is received before the table map event, and as you point out, it is not associated to a table-specific event, so it is not filtered out.  I have written up [NIFI-3730](https://issues.apache.org/jira/browse/NIFI-3730) to address this.